Configurable Zuul X-Forwarded-Account header value

lhazlewood opened the pull request:
Resolves #1073
Resolves #962

To review, please *don't* just look at the Pull Request side-by-side diff.  The docs need to be reviewed in rendered form to get a feel for what a customer sees (to see if it makes sense and is usable).

To review:

Check out the pull request as a new branch:

    $ git fetch origin pull/1176/head:pr-1176

Then switch to this new branch:

    $ git checkout pr-1176

Then generate and open/read the docs:

    $ cd docs
    $ ./build.sh sczuul # you shouldn't see any errors or warnings here
    $ open build/sczuul/build/html/index.html

Please open the `Forwarded Request` link on the side.  This is the docs page that was edited heavily.

Please look for:

- Anything that doesn't make sense or is confusing
- Is the page layout/menu correct?  Are topics introduced in a correct/understandable manner?

Please add any improvement suggestions to the PR review.

When you're done, you can delete that local pr branch:

    $ git checkout master
    $ git branch -D pr-1176

lhazlewood commented:
To review, please *don't* just look at the Pull Request side-by-side diff.  The docs need to be reviewed in rendered form to get a feel for what a customer sees (to see if it makes sense and is usable).

To review:

Check out the pull request as a new branch:

    $ git fetch origin pull/1176/head:pr-1176

Then switch to this new branch:

    $ git checkout pr-1176

Then generate and open/read the docs:

    $ cd docs
    $ ./build.sh sczuul # you shouldn't see any errors or warnings here
    $ open build/sczuul/build/html/index.html

Please open the `Forwarded Request` link on the side.  This is the docs page that was edited heavily.

Please look for:

- Anything that doesn't make sense or is confusing
- Is the page layout/menu correct?  Are topics introduced in a correct/understandable manner?

Please add any improvement suggestions to the PR review.

When you're done, you can delete that local pr branch:

    $ git branch -D pr-1176

mrioan commented on docs/source/forwarded-request.rst:
```diff
@@ -40,150 +40,965 @@ If you change this name, ensure the origin server(s) behind your gateway know to
+         header:
+           jwt:
+             key:
+               kid: my signing key
```
kid: my signing key id

mrioan commented on docs/source/forwarded-request.rst:
```diff
@@ -40,150 +40,965 @@ If you change this name, ensure the origin server(s) behind your gateway know to
+
+2.  If you need access to the HttpServletRequest/Response during the account-to-string conversion process, you can
+    define your own :ref:`stormpathForwardedAccountHeaderValueResolver` bean.
+
```
`In either case you will need to add the proper bean in your Zuul application. Remember that we are talking about a 2-step process here: 1) we first need to convert the Account (in the gateway) and 2) you can then read it in your origin server. If you want to customize the way the account is converted then you need to do so in step 1.`

lhazlewood replied:
simplified this a bit and moved it to a note/admonition for better awareness.

lhazlewood commented on docs/source/forwarded-request.rst:
```diff
@@ -40,150 +40,965 @@ If you change this name, ensure the origin server(s) behind your gateway know to
+your rules, and the resulting output will be a single JSON document that has the same graph structure as your
+specified conversion graph.
+
+So what does conversion configuration look like?  Without jumping too far ahead, here is the default Account
```
Good idea!  Thanks :)

lhazlewood commented on docs/source/forwarded-request.rst:
```diff
@@ -40,150 +40,965 @@ If you change this name, ensure the origin server(s) behind your gateway know to
+       account:
+         header:
+           value:
+             strategy: scalars
```
Should what be `json`?  The `strategy` value?

If you mean the strategy value, the answer is no.  Unless strategy is `single`, every other strategy means JSON, so just `json` isn't descriptive enough.

Does this make sense?  Do you have any recommendations here? (Feedback welcome!)

mrioan replied:
Yes, I was confused at first. You were introducing the JSON value and then you put `scalars` in the strategy. I think I was expecting a data structure (or data representation) name there (like JSON). But the strategy does not define structure, the strategy defines algorithms/factories/functions. I was not sure what a strategy was. Maybe you should introduce what a strategy conceptually is. I think that would help

lhazlewood replied:
Sorry - given that I've made changes to the PR before seeing this comment, I'm still confused as to what part of the page you were commenting on.  Could you please remind me?

FWIW, strategy here is in the spirit of the [strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern), where the name is an 'id' of the strategy/algorithm to use.  Those strategies are fully defined in the `strategy` section of the document, so I'm confused as to where else you think we might need documentation.

Suggestions welcome.  Thanks!

jhericks replied:
Earlier, in the Single value section, you say, "This configuration states that JWT is disabled, and we’ll use a value conversion strategy of `single` which means we want the header to be a single account field value (we’ll talk about JWTs and conversion strategies later)." Things are appropriately hyperlinked there, but it might easy to blow past.  It might be as simple as hyperlinking more generously as you mention strategies.  I have had an introduction to `strategy: scalars` but without that introduction, it's very jargony.

lhazlewood commented on docs/source/forwarded-request.rst:
```diff
@@ -40,150 +40,965 @@ If you change this name, ensure the origin server(s) behind your gateway know to
+         - don't include the account's ``href`` field.  My origin server won't talk directly with Stormpath and won't
+           know what to do with that url, so exclude it
+
+         - ``customData`` isn't a scalar, but I want it included anyway, so I'm going to define conversion rules for it
```
I stepped through the English description one by one, and I cannot see where there are differences between what is written and what is reflected in the configuration.

I did however clean up the indentation in the English, so hopefully that will show the 1:1 correlation.

If this still isn't clear, or I'm missing something, please let me know!

mrioan replied:
@lhazlewood First of all I will write as a user who is reading things sequentially. He did not read the available strategies yet (as they are further below of this particular section)

Let me explain what I see then...

You say this:
```
*`customData` isn’t a scalar, but I want it included anyway, so I’m going to define conversion rules for it too. Those are:
```

And the configuration is:

```
    customData:
        strategy: scalars
            fields:
                href:
                  enabled: false
```

For me yhis is not 1:1 (P is for Property, L is for human language):

P1: `customData` isn’t a scalar, but I want it included anyway: 
L1:  the *BUT* is not clear. Me as a user do not knot that non-scalars are not included in the scalars strategy (I can infer because of the strategy name, but I am not really sure)

This means that there is some default logic that you know abut the strategy but I don't. So, I would write:


``` 
`customData` isn’t a scalar (this means that the `scalars` strategy will skip this field altogether). I still want particular intra-CustomData fields to be added to my resulting JSON, so I'm going to define conversion rules for it
```

Makes sense? Maybe I am being too literal... Don't worry if you think otherwise!

mrioan commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ElementsConversion.java:
```diff
@@ -0,0 +1,101 @@
+package com.stormpath.sdk.convert;
```
Missing header

lhazlewood replied:
done

mrioan commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/Conversions.java:
```diff
@@ -0,0 +1,60 @@
+package com.stormpath.sdk.convert;
```
Missing header

lhazlewood replied:
done

mrioan commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/MapToJwtConverter.java:
```diff
@@ -0,0 +1,229 @@
+package com.stormpath.sdk.convert;
```
Missing header

lhazlewood replied:
done

mrioan commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ResourceConverter.java:
```diff
@@ -0,0 +1,284 @@
+package com.stormpath.sdk.convert;
```
Missing header

lhazlewood replied:
done

mrioan commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ResourceConverter.java:
```diff
@@ -0,0 +1,284 @@
+
+    private static final Logger log = LoggerFactory.getLogger(ResourceConverter.class);
+
+    public static final Conversion DEFAULT_CONFIG = new Conversion()
```
I think you can use `Conversions.withField` instad of `new Conversion()`

lhazlewood replied:
doesn't really get us anything, but I changed it anyway :)

mrioan commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ResourceConverter.java:
```diff
@@ -0,0 +1,284 @@
+     * @param config the {@code Conversion} that indicates how to convert the function argument into an output value.
+     */
+    public void setConfig(Conversion config) {
+        this.config = config;
```
`assert.NotNull` ?

lhazlewood replied:
added

mrioan commented on extensions/servlet/src/test/groovy/com/stormpath/sdk/convert/ConversionTest.groovy:
```diff
@@ -0,0 +1,51 @@
+package com.stormpath.sdk.convert
```
Missing header

lhazlewood replied:
done

mrioan commented on extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/zuul/config/JwtConfig.java:
```diff
@@ -0,0 +1,63 @@
+package com.stormpath.spring.cloud.zuul.config;
```
Missing header

lhazlewood replied:
done

mrioan commented on extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/zuul/config/StormpathZuulAccountHeaderConfig.java:
```diff
@@ -0,0 +1,51 @@
+package com.stormpath.spring.cloud.zuul.config;
```
Missing header

lhazlewood replied:
done

lhazlewood commented on api/src/main/java/com/stormpath/sdk/lang/Instants.java:
```diff
@@ -30,9 +31,33 @@
+     *
      * @param time the UTC instant to convert, represented as the number of milliseconds that have elapsed since midnight, January 1, 1970
-     * @param to the target {@link TimeZone} for the conversion
+     * @param to   the target {@link TimeZone} for the conversion
```
That's intentional - at least as IntelliJ formats it.  All param descriptions are left justified with each other.

lhazlewood commented on api/src/main/java/com/stormpath/sdk/lang/Instants.java:
```diff
@@ -51,9 +77,10 @@ public static long convertDateToUTC(long time, TimeZone from) {
      * @param time the instant to convert, represented as the number of milliseconds that have elapsed since midnight, January 1, 1970 in the {@code from} {@link TimeZone}
      * @param from the original {@link TimeZone}
-     * @param to the target {@link TimeZone} for the conversion
+     * @param to   the target {@link TimeZone} for the conversion
```
IntelliJ ;)

lhazlewood commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/Conversion.java:
```diff
@@ -0,0 +1,226 @@
+     * @return this object for method chaining.
+     */
+    public Conversion setName(String name) {
+        this.name = name;
```
Not in this case - null is valid (it just means use the default name)

lhazlewood commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/Conversion.java:
```diff
@@ -0,0 +1,226 @@
+     * @see ConversionStrategyName#SINGLE
+     */
+    public Conversion setField(String field) {
+        this.field = field;
```
same

lhazlewood commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ConversionStrategyName.java:
```diff
@@ -0,0 +1,55 @@
+public enum ConversionStrategyName {
+
+    /**
+     * A strategy name that indicates only fields explicitly defined in the
```
done

lhazlewood commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ConversionStrategyName.java:
```diff
@@ -0,0 +1,55 @@
+    DEFINED,
+
+    /**
+     * A strategy name that indicates that the conversion output should be just one of the source object's field values.
```
done

lhazlewood commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/ConversionStrategyName.java:
```diff
@@ -0,0 +1,55 @@
+
+    /**
+     * A strategy that represents <em>ALL</em> fields of the source object should be in the output.  Be careful
+     * when choosing this strategy as the output could be sufficiently larger than desired.
```
done

lhazlewood commented on extensions/servlet/src/main/java/com/stormpath/sdk/convert/MapToJwtConverter.java:
```diff
@@ -0,0 +1,229 @@
+ *
+ * @since 1.3.0
+ */
+public class MapToJwtConverter implements Function<Map<String, ?>, String> {
```
agreed.

mrioan commented on docs/source/forwarded-request.rst:
```diff
@@ -31,159 +31,1108 @@ If you want to specify a different header name, set the ``stormpath.zuul.account
+             enabled: false
+
+This ensures the header value is *NOT* a JWT, but either an :ref:`Account JSON document <forwarded account json>` or a
+ :ref:`single string value <forwarded account single field>` as documented above, but beware of the security implications.
```
This line seems to be aligned to the center of the page rather than to the left

mrioan commented on docs/source/forwarded-request.rst:
```diff
@@ -31,159 +31,1108 @@ If you want to specify a different header name, set the ``stormpath.zuul.account
+         - ``groups`` isn't a scalar (it's a ``GroupsCollection`` object), but I want it included anyway.
+
+           - However, in this case I want to include *only* fields that are explicitly *defined** in its ``fields``
+             list.  (In this case, even though the strategy is ``defined``, no actual ``fields`` have been specified.
```
It looks wrong to me that this long sentence is between parentheses (ok if intended, yo know better english than me ;)) . Additionally I think the last period for a inside a parenthesis is wrong (again, ok if intended, yo know better english than me ;)

lhazlewood replied:
Yeah, the parenthesis here indicates an aside - a side comment that underscores/justifies the previous explanation.  It's totally debatable as to whether or not this is 'proper' or not, but I think it helps in tone so I guess I'll leave it in unless someone is confused by it.

jhericks commented:
In your explanatory example under the **Account JSON** section, you say, "Include any of the customData’s scalar (i.e. non-object/collection) properties automatically."  That little parenthetical explanation of the what we mean by 'scalar' would be helpful even earlier, the first time it's mentioned.

jhericks commented:
In *Caution* box under **list** explanation, you say, "We typically strongly recommend that you DO NOT use the list strategy if forwards-compatibility is important to you - JSON arrays are inflexible and cannot support additional properties over time, whereas JSON objects are flexible and allows for future property expansion."

'forwards-compatibility' should be 'forward-compatibility and you should make the 'objects' and 'are' and 'allows' agree on plural vs. singular.

jhericks commented:
In the **Account JSON** section, where you start with, "Now that we’ve seen a good example, let’s cover all the possible conversion config properties to explain their functionality," it could use a heading or something because when I went back to find it it was hard to scan for.  Also, I see that you listed each conversion config property in alphabetical order, but I feel like the `strategy` config explanation gives good context for some of the other config properties.

jhericks commented:
Looks like some formatting issues in the `header` examples under the **JWT** section.

jhericks commented:
The fact that `valueClaim` is a heading over `enabled` and `name` is subtle.  I missed it and it made the whole section read very confusingly until I figured that out.

lhazlewood commented:
@jhericks 

Thanks for the review!

Results:

- scalar explanation: we're on the same page. While you were reviewing an earlier version of this PR, I was making this change.  I now link to a page that explains what scalars are w.r.t. programming.
- **list** explanation: edited, thanks!
- conversion rules: added a section header to make this more noticeable.
- conversion config alphabetical order: I added a sentence before this auto-rendered table of contents that explicitly calls out that the `strategy` sub section may be particularly helpful.
- `header` examples formatting: fixed
- value claim heading:  I'm not sure I can do anything more here from a content perspective - I'm using standard Sphinx section headers to indicate `name` and `enabled` are proper children sections of `valueClaim`.  Our theme is responsible for rendering them in a way that makes them stand out (or not) appropriately.  Any thoughts or suggestions on non-theme-related improvements?
